src/device/intrinsics/math.jl

```diff
@@ -103,17 +102,98 @@ end
 
 @device_override Base.log(x::Float64) = ccall("extern __nv_log", llvmcall, Cdouble, (Cdouble,), x)
 @device_override Base.log(x::Float32) = ccall("extern __nv_logf", llvmcall, Cfloat, (Cfloat,), x)
+@device_override function Base.log(x::Float16)
+    log_x = @asmcall("""{.reg.b32        f, C;
+                  .reg.b16        r,h;
+                  mov.b16         h,\$1;
+                  cvt.f32.f16     f,h;
+                  lg2.approx.ftz.f32  f,f;
```

Comment:
Seeing `.approx` here; are these the fastmath versions?

Reply:
It's from their own `log` implementation which is *not* fast math (https://github.com/cupy/cupy/blob/620183256d25eb463081a8bac2a7a965d35db66b/cupy/_core/include/cupy/_cuda/cuda-12/cuda_fp16.hpp#L2478), I guess they use the approximate method for fp32 and assume it doesn't hurt fp16 accuracy too much?